### PR TITLE
uichange Vending Machine

### PR DESCRIPTION
**Pull request by DEATHB4DEFEAT**
## About the PR

**Media**

image
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: DEATHB4DEFEAT
- tweak: Vending machines are good now.

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
I don't really think having the side portion with the image, name, description and vending button really adds much. Having to click the vending button seems very clunky, all of the info given there is already present (minus the useless description), and it creates a really weird dead space in the bottom right corner which is unappealing to look at.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 1]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
         {
-            MinSize = SetSize = (250, 150);
             RobustXamlLoader.Load(this);
             IoCManager.InjectDependencies(this);
 
             VendingContents.OnItemSelected += args =>
             {
-                OnItemSelected?.Invoke(args);
+                // OnItemSelected?.Invoke(args);
+                onItemSelected(args);
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:30]: C# style guide, also dummy comment?

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 2]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
+                foreach (var item in VendingContents)
+                {
+                    if (item.Selected)
+                    {
+                        finalPos = pos;
+                        pos++;
+                        break;
+                    }
+                    pos++;
+                }
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:48]: You can just update this when the item is selected...

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 3]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
+
+                if (finalPos != 0 && finalPos < pos) OnItemSelected?.Invoke(new ItemList.ItemListSelectedEventArgs(finalPos, VendingContents));
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:50]: Please don't inline code like this.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 4]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
+            };
+
+
+            // Placeholder VendingInfo
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            var ent = entityManager.SpawnEntity("VendingMachineCola", MapCoordinates.Nullspace);
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:56]: Any reason you don't use a spriteview of the entity?

You also never cleaned this entity up.

**Comment by moonheart08** [comment]
This could be a lot more compact.

**Comment by metalgearsloth** [comment]
> I don't really think having the side portion with the image, name, description and vending button really adds much. Having to click the vending button seems very clunky, all of the info given there is already present (minus the useless description), and it creates a really weird dead space in the bottom right corner which is unappealing to look at.

Having the larger spriteview is nice but the right-side should be set to vertical expand and shorter.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 5]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
+            VendingInfo.AddChild(new Control { MinSize = (10, 10) });
+
+            message = new FormattedMessage();
+            message.AddMarkup("Select an item from the left to preview it.");
+            var VendingDescription = new RichTextLabel();
+            VendingDescription.SetMessage(message);
+            VendingInfo.AddChild(VendingDescription);
+
+            VendButton.Disabled = true;
+            // Placeholder VendingInfo
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:82]: Dummy comment.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 6]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
         }
 
+        private void onItemSelected(ItemList.ItemListSelectedEventArgs args)
+        {
+            ItemList.Item selected = VendingContents[args.ItemIndex];
+            VendingMachineInventoryEntry? entry = (VendingMachineInventoryEntry?) selected.Metadata!;
+
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:89]: This looks like a dirty null suppression, it should be handled properly.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 7]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
+            VendingInfo.Children.Clear();
+
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            var ent = entityManager.SpawnEntity(entry.ID ?? "VendingMachineCola", MapCoordinates.Nullspace);
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:93]: As above.

Also just store the IEntityManager dependency.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 8]
```diff
@@ -18,16 +21,109 @@ public sealed partial class VendingMachineMenu : DefaultWindow
...
+
+            var metaData = entityManager.GetComponent<MetaDataComponent>(ent);
+            var sprite = entityManager.GetComponent<SpriteComponent>(ent);
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:96]: TryGet spritecomponent.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs [hunk 9]
```diff
@@ -72,17 +168,17 @@ public void Populate(List<VendingMachineInventoryEntry> inventory)
                 if (itemName.Length > longestEntry.Length)
                     longestEntry = itemName;
 
-                vendingItem.Text = $"{itemName} [{entry.Amount}]";
+                vendingItem.Text = $" [{entry.Amount}] {itemName}";
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:171]: Use the margin rather than whitespace in text.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Needs localising.

**Comment by moonheart08** [comment]
I think the extra work required to vend isn't really a benefit, the item desc/view is fine.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml [hunk 10]
```diff
@@ -1,6 +1,26 @@
...
+    MinSize="650 250"
+    SetSize="675 250">
+    <BoxContainer Orientation="Horizontal">
+        <!-- List of vendable items -->
+        <ItemList Name="VendingContents" MinWidth="350" SizeFlagsStretchRatio="8" HorizontalExpand="True" VerticalExpand="True" />
+
+        <Control MinSize="10 10" />
+
+        <BoxContainer Orientation="Vertical" MinWidth="275" MaxWidth="275" SizeFlagsStretchRatio="2" HorizontalExpand="True" VerticalExpand="True">
+            <Button Disabled="True">
```
Comment (metalgearsloth) [inline Content.Client/VendingMachines/UI/VendingMachineMenu.xaml:11]: I went back and realised I forget to check if this is used, please don't hack buttons like this.

File: Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs (hunk 4 above)
Comment (DEATHB4DEFEAT) [reply Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:56]: it is using spriteview?

Comment (DEATHB4DEFEAT) [reply Content.Client/VendingMachines/UI/VendingMachineMenu.xaml.cs:56]: it is using spriteview

**Comment by jamessimo** [comment]
Ha, you and I seem to be on a similar wave length! https://discord.com/channels/310555209753690112/346743671787356160/1074236362984804393

This was my rough design I made a few weeks back. 
Vending Machine

But if you check the discord there was some push back and I put this on the bottom of my list until I could get a design with a better consensus. I recommend you share your UI changes there and get some feedback before making a PR. Its nice to get end-users feedback as well as devs! 

Look me up on discord and would love to share my other improvements and discuss more!

**Comment by jamessimo** [comment]
Is the view of the machine the same for all vending machines? I see the drinks machine but your in a snacks machine?

**Comment by DEATHB4DEFEAT** [comment]
placeholder icon

**Comment by metalgearsloth** [comment]
@DEATHB4DEFEAT are you coming back to this.

**Comment by DEATHB4DEFEAT** [comment]
> @DEATHB4DEFEAT are you coming back to this.

keep forgetting to do so, I'll try to get something done later today

**Comment by metalgearsloth** [comment]
Re-open when you get a chance to work on it.

**Pull request closed without merging**